### PR TITLE
Experiment: Generic ElasticSearch query

### PR DESCRIPTION
**Pull request by rsevilla87**
## Type of change

- [ ] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Adding a new `generic_search` method to retrieve data from arbitrary queries, arbitrary buckets and arbitrary aggregations. Example of result after running a testing script against our aws dev instance

https://gist.github.com/rsevilla87/ea63357d1243e1614a2c276eb09c032c

Some benefits:

- Just one query get data broken down by a metadata field.
- Permits nested buckets/metrics, they can be specified with a dot notation, example `metadata.ocpMajorVersion.keyword`

It's major downside can be the format of the data returned, it's a single JSON document with lots of nested fields, (one per term/metric aggregation. Good thing is that the value of the metric aggregation is returned at the lowest level and the key names are predictable, in this case:

```json
 "aggregations": {
    "metadata.platform.keyword": {
      "buckets": [
        {
          "key": "BareMetal",
          "metadata.sdnType.keyword": {
            "buckets": [
              {
                "key": "OVNKubernetes",
                "metadata.ocpMajorVersion.keyword": {
                  "buckets": [
                    {
                      "key": "4.14",
                      "avg-latency": {
                        "value": 2260.4151167980954
                      }
                    },
                    {
                      "key": "4.13",
                      "avg-latency": {
                        "value": 11727.016764266851
                      }
                    }
                  ]
                }
              }
            ]
          }
        },
```

Thoughts??

### CODE REVIEW

**Comment by rsevilla87** [comment]
cc: @shashank-boyapally @jtaleric

File: fmatch/matcher.py [hunk 1]
```diff
@@ -7,12 +7,12 @@
 import logging
 
 # pylint: disable=import-error
-from elasticsearch import Elasticsearch
+from elasticsearch7 import Elasticsearch
```
Comment (jtaleric) [inline fmatch/matcher.py:10]: curious about this change?

Comment (shashank-boyapally) [reply fmatch/matcher.py:10]: I think `elasticsearch7` can be installed using `elasticsearch==7.x.x`, this has been a previous issue for me too.

Comment (rsevilla87) [reply fmatch/matcher.py:10]: yeah, if you don't specify the module version, the python interpreter just picks the latest one available

File: fmatch/matcher.py [hunk 2]
```diff
@@ -38,6 +38,39 @@ def __init__(
...
+        """
+        s = Search(using=self.es, index=self.index)
+        q = None
+        for query in queries:
+            if not q:
+                q = Q("match", **(query))
+            else:
+                q = q & Q("match", **query)
+        s.query = q
+        x = None
```
Comment (vishnuchalla) [inline fmatch/matcher.py:61]:
```suggestion
        x = None
        for bucket in buckets:
            a = A("terms", field=bucket)
            if x is None:
                x = s.aggs.bucket(bucket, a)
            else:
                x = x.bucket(bucket, a)
        x.bucket(metric_name, A(agg_type, field=field))
```

File: fmatch/matcher.py [hunk 3]
```diff
@@ -38,6 +38,39 @@ def __init__(
         self.es = Elasticsearch([self.es_url], timeout=30)
         self.data = None
 
+    def generic_search(self, queries: list, buckets: list, metric_name: str, agg_type: str, field: str):
```
Comment (vishnuchalla) [inline fmatch/matcher.py:41]: Overall I like this idea. Should be useful in reducing multiple queries to one on a single ES index atleast.

**Pull request closed without merging**